Feature/support workspace properties

nixel2007 opened the pull request:
To make the maintainer's life easier, I...

- [x] created at least 1 spec to cover my changes,
- [x] `npm test`ed my code and it's green like an :green_apple:,
- [x] adjusted the `readme.md`, if it was necessary and
- [x] would like to receive a virtual :beer: after that hard work!

The main idea is setting "Workspace" package config when active pane is changed (cause https://github.com/atom/whitespace/issues/104 is still not implemented).
Also I've fixed an error message with promise, when opening settings tab.

nixel2007 commented:
Link to https://github.com/sindresorhus/atom-editorconfig/issues/90

nixel2007 commented:
Maybe it will be good idea to store the default values for whitespace-package settings. I think, editorconfig should restore this settings to default if it can't find `.editorconfig` file or there is no entry for this setting.

florianb commented:
Hi @nixel2007 -- thank you very much for your contribution. 😄 🍺

I took a deeper look on your proposed changes and have one point to discuss and two questions. I would really appreciate if you could help me getting through this.

I am working on implementing `trim_trailing_whitespaces` and `insert_final_newline`, too and i didn't consider changing the `whitespace.config` for the following reason:

I think Atom-Editorconfig should not change any config-settings of the editor-environment in general for two reasons. Other packages may observe these config-settings and may take action even though the user hasn't changed anything. But more important, programmatically changed config-options immediately persisted to the `config.cson` -- tying this up after an editor gets closed makes these things even more complicated. Besides of that do any config-changes affect all open editors and workspaces.

Currently i think implementing `ttw` and `ifn` via config-settings wouldn't play well with the atom-environment. And i can't imagine a robust strategy how to "revert" any config-changes properly.

What do you think?

You also added support for workspaces, i would like to ask why you did that - did anything not work properly using panes?

And you mentioned a error message when opening the config-tab -- could you give me some more information about that?

nixel2007 commented:
hi @florianb!

> Currently i think implementing ttw and ifn via config-settings wouldn't play well with the atom-environment. And i can't imagine a robust strategy how to "revert" any config-changes properly.

Yes, this is the worst thing about ttw and ifn properties. But i have some ideas about reverting.
First off all - you can store whitespace settings when package is activated.
Then we have 3 options:
* user opened a file /w .editorconfig - we load editorconfig settings as we do now
* user opened a file /wo .editorconfig - we load stored values
* user close editor - we load stored values

All this options can be observed via atom events.
The only bad thing in this scenario - user can manualy change the whitespace setting.

The second idea is changing the `config.cson` file instead of setting the whitespace settings. I didn't think about it a lot, but this can be a workaround solution.

> You also added support for workspaces, i would like to ask why you did that - did anything not work properly using panes?

Workspaces... do you mean the `setWorkspaceConfig` function? I've added this function cause `editor` settings is binded to specific editor (pane) with specific file, but changing `whitespace` properties is something about overall atom worspace setting.
And the second idea is - you don't need to reconfig editor (pane) options, when you change the tab (onDidChangeActivePaneItem event). you only need to reconfig whitespace settings.

> And you mentioned a error message when opening the config-tab -- could you give me some more information about that?

Checkout master branch, open project with .editorconfig settings (to make sure package is activated), open chome console (dev tools) and open Settings tab. You'll see an error message about rejected promise without a rejection handling.

nixel2007 commented:
Ah, i see. I wrote `The main idea is setting "Workspace" package` on the top. I mean the `whitespace` package of course.

nixel2007 commented:
And we have a third solution... Implement https://github.com/atom/whitespace/issues/104 :D

sindresorhus commented:
> I think Atom-Editorconfig should not change any config-settings of the editor-environment in general for two reasons. Other packages may observe these config-settings and may take action even though the user hasn't changed anything. But more important, programmatically changed config-options immediately persisted to the config.cson -- tying this up after an editor gets closed makes these things even more complicated. Besides of that do any config-changes affect all open editors and workspaces.

I also don't think this is a good idea. Would rather see https://github.com/atom/whitespace/issues/104 happen.

Arcanemagus commented:
Just as another point as to why changing that setting is a bad idea: What happens if this package changes the setting, and then Atom crashes without letting the cleanup/revert task run? No software is perfect so you should assume it will break at some point 😛.

florianb commented:
💭 I guess whitespace is just not properly handling the per-language-scope and that's Atom's preferred way to handle (also inline) grammar-specific settings. However, that paradigm is completely different to Editorconfig's per file settings and thats why we have troubles making this package editorconfig-conform.

To get back to the PR, at this moment i wouldn't add any config-based functionality. What benefit will we have from adding the `onDidChangeActivePaneItem `-hook?

I currently had no time testing the setting-page issue. Please be patient. ☕

sindresorhus commented:
I much prefer Sublime's config handling where you can change anything on a per buffer basis. Atom really needs that.

florianb commented:
I can understand the decision to put the configuration-task aside the text-editing tasks, but i think in any case trimming trailing whitespaces and inserting a final newline should be considered being a textbuffer-function since those are widely adopted first class functions of a texteditor.

florianb commented:
Since there were no new activities for some days i close this merge request. I am sorry that we didn't adopt your work.

nixel2007 commented:
@florianb no problem.
I've put this on my ToDo-list. When i'll get some free time, i'll try to rethink the concept and/or implement the whitespace-package part..
